docs/snapshotting/versioning.md

```diff
@@ -61,59 +35,37 @@ emulation, KVM and vCPUs) with 2 exceptions - serial emulation and vsock backend
 
-| Field | Bits| Description |
-|----|----|----|
-| magic_id | 64 | Firecracker snapshot, architecture (x86_64/aarch64) and storage version. |
-| version  | 16 | The snapshot version number internally mapped 1:1 to a specific Firecracker version. |
+| Field | Bits | Description |
+|-------|------|-------------|
+| magic_id | 64 | Firecracker snapshot and architecture (x86_64/aarch64). |
+| version | M | The snapshot data format version (`MAJOR.MINOR.PATCH`) |
 | state | N | Bincode blob containing the microVM state. |
-| crc| 64 | Optional CRC64 sum of magic_id, version and state fields. |
-
-**Note**: the last 16 bits of `magic_id` encode the storage version which specifies
-the encoding used for the `version` and `state` fields. The current
-implementation sets this field to 1, which identifies it as a [Serde bincode](https://github.com/servo/bincode)
-compatible encoder/decoder.
-
-### Version tolerant ser/de
-
-Firecracker reads and writes the `state` blob of the snapshot by using per
-version, separate serialization and deserialization logic. This logic is mostly
-autogenerated by a Rust procedural macro based on `struct` and `enum`
-annotations. Basically, one can say that these structures support versioning.
-The versioning logic is generated by parsing a structure's history log (encoded
-using Rust annotations) and emitting Rust code.
-
-Versioned serialization and deserialization is divided into two translation layers:
-
-* field translator,
-* semantic translator.
-
-The _field translator_ implements the logic to convert between different
-versions of the same Rust POD structure: it can deserialize or serialize from
-source version to target.
-The translation is done field by field - the common fields are copied from
-source to target, and the fields that are unique to the target are
-(de)serialized with their default values.
-
-The _semantic translator_ is only concerned with translating the semantics of
-the serialized/deserialized fields.
-
-The _field translator_ is generated automatically through a procedural macro,
-and the _semantic translation methods_ have to be annotated in the structure
-by the user.
-
-This block diagram illustrates the concept:
-
-![Versionize](
-../images/versionize.png?raw=true
-"Versionize layers")
+| crc | 64 | Optional CRC64 sum of magic_id, version and state fields. |
+
+The snapshot format has its own version encoded in the snapshot file itself
+after the snapshot's `magic_id`. The snapshot format version is independent of
+the Firecracker version and it follows semantic versioning, i.e.
+`MAJOR.MINOR.PATCH`. That means that backwards incompatible changes in the data
+format will cause `MAJOR` version number bump, whereas backwards compatible
+changes will cause a `MINOR` version number bump.
+
+For example, that means that a Firecracker binary that supports snapshot data
+format `x.y.z` will be able to load all snapshot files with version `x.Y.z`
+where `Y <= y`.
```

Comment:
Wait, is this really semantic versioning? Shouldn't a Firecracker binary that declares support for snapshot version format, say, 2.1.0 also be able to load version, say, 2.5.0?

Also, this should say something like "will be able to load all snapshot files with version `x.Y.Z` with `Y<y` or `Y=y and Z<=z`.

Reply:
I removed mentions of semantic versioning. In any case, with bincode we will not be able to make any backwards compatible changes to the snapshot format. If in the future we do, we should define this properly, for the moment we just mention the version format `MAJOR.MINOR.PATCH` which is generic enough to describe whatever backwards compatibility we need.

For the record, SchemaVer (which is based on semantic versioning) seems to fit well to what we have in mind: https://snowplow.io/blog/introducing-schemaver-for-semantic-versioning-of-schemas/

Reply:
Sounds good!